### PR TITLE
Chapter 1 solutions

### PR DESCRIPTION
**Pull request by atlas17**
### Solutions for Chapter 1

Hello there! :)

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter1.hs [hunk 1]
```diff
@@ -416,7 +416,7 @@ task is to specify the type of this function.
 >>> squareSum 3 4
 49
 -}
-
+squareSum :: Num a => a -> a -> a
```
Comment (chshersh) [inline src/Chapter1.hs:419]: We explain `Num` in the following chapters, so you are one step ahead 😸

In here `Int -> Int -> Int` would also work, as a more specific type 🙂

File: src/Chapter1.hs [hunk 2]
```diff
@@ -477,8 +477,8 @@ Implement a function that returns the last digit of a given number.
   results. Or you can try to guess the function name, search for it and check
   whether it works for you!
 -}
--- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-lastDigit n = error "lastDigit: Not implemented!"
+lastDigit :: Integral a => a -> a
+lastDigit n = mod n 10
```
Comment (chshersh) [inline src/Chapter1.hs:481]: Your implementation is almost correct :ok: 
Unfortunately, it returns negative numbers on negative inputs because of how `mod` works. Sometimes corner cases can be tricky to spot and fix...

File: src/Chapter1.hs [hunk 3]
```diff
@@ -556,7 +559,8 @@ True
 >>> isVowel 'x'
 False
 -}
-isVowel c = error "isVowel: not implemented!"
+isVowel :: Char -> Bool
+isVowel c = c `elem` "aeouiAEOUI"
```
Comment (chshersh) [inline src/Chapter1.hs:563]: Absolutely correct! 👍🏼 

One note in here, that sometimes, `elem` could be slower than the explicit pattern matching. I remember there were some benchmarks on one particular case, that showed how moving to pattern matching on each case separately drastically decrease time 🐎

File: src/Chapter1.hs [hunk 4]
```diff
@@ -619,8 +623,8 @@ Implement a function that returns the sum of the last two digits of a number.
 Try to introduce variables in this task (either with let-in or where) to avoid
 specifying complex expressions.
 -}
-
-sumLast2 n = error "sumLast2: Not implemented!"
+sumLast2 :: Integral a => a -> a
+sumLast2 n = mod n 10 + div (mod n 100) 10
```
Comment (chshersh) [inline src/Chapter1.hs:627]: That is a wonderful solution! 👏🏼  You correctly noticed that it is the `div` and `mod`, cool 😎 

One hint to make your solution even shorter: the standard library has the [`divMod`](https://hackage.haskell.org/package/base-4.14.0.0/docs/Prelude.html#v:divMod) function, that is actually combines inside both `div` and `mod`. And this is exactly what you use!.

So you could write it this way:

```haskell
(first, second) = divMod (mod x 100) 10
```

You can see how we could pattern match on the pair 🙂 

It also looks like this function also should handle negative numbers better :thinking: 
Sometimes it's tricky to think about all corner cases and handle them properly...

**Review by chshersh — APPROVED** [review]
Congrats on finishing Chapter One!
And well done with the Advanced Task :medal_sports: 

I left some comments and general feedback as a part of my review :+1:

File: src/Chapter1.hs (hunk 2 above)
Comment (atlas17) [reply src/Chapter1.hs:481]: Yes, it seems I forgot to handle corner cases every time they were not mentioned in the task and not present in tests.

File: src/Chapter1.hs (hunk 3 above)
Comment (atlas17) [reply src/Chapter1.hs:563]: Do you mean one guard for every vowel letter? Or something like `| c == 'a' || c == 'e' ...` ?

File: src/Chapter1.hs (hunk 4 above)
Comment (atlas17) [reply src/Chapter1.hs:627]: Thank you for the review!

File: src/Chapter1.hs (hunk 3 above)
Comment (chshersh) [reply src/Chapter1.hs:563]: I guess both solutions are equivalent :+1:
